service/router/basic.go

```diff
@@ -18,225 +18,228 @@ package router
 
 import (
 	"context"
-	"fmt"
 	"reflect"
 	"sort"
 )
 
-// Handler contains middlewares and executor.
-type Handler struct {
-	Middlewares []Middleware
-	Executor    Executor
+// handler contains middlewares and executor.
+type handler struct {
+	middlewares []Middleware
+	inspector   Inspector
 }
 
 // AddMiddleware adds middleware to the router node.
 // If the router matches a path, all middlewares in the router
 // will be executed by the returned executor.
-func (h *Handler) AddMiddleware(ms ...Middleware) {
-	h.Middlewares = append(h.Middlewares, ms...)
+func (h *handler) AddMiddleware(ms ...Middleware) {
+	h.middlewares = append(h.middlewares, ms...)
+}
+
+// Middlewares returns all middlewares of the router.
+// Don't modify the returned values.
+func (h *handler) Middlewares() []Middleware {
+	return h.middlewares
 }
 
 // AddExecutor adds executor to the router node.
 // A router can hold many executors, but there is only one executor
 // is selected for a match.
-func (h *Handler) AddExecutor(es ...Executor) {
-	var executors Executors
-	if h.Executor != nil {
-		if array, ok := h.Executor.(Executors); ok {
-			executors = array
-		} else {
-			executors = Executors{h.Executor}
-		}
-	}
-	for _, e := range es {
-		if e != nil {
-			if array, ok := e.(Executors); ok {
-				executors = append(executors, array...)
-			} else {
-				executors = append(executors, e)
-			}
-		}
-	}
-	if executors != nil {
-		h.Executor = executors
-	}
+func (h *handler) SetInspector(inspector Inspector) {
+	h.inspector = inspector
+}
+
+// Inspector gets inspector from the router node.
+// Don't modify the returned values.
+func (h *handler) Inspector() Inspector {
+	return h.inspector
 }
 
 // Merge merges middlewares and executors.
-func (h *Handler) Merge(o *Handler) {
-	h.AddMiddleware(o.Middlewares...)
-	h.AddExecutor(o.Executor)
+func (h *handler) Merge(o *handler) error {
+	h.AddMiddleware(o.middlewares...)
+	if h.inspector != nil {
+		if o.inspector != nil {
+			return ConflictInspectors.Error()
+		}
+	} else {
+		h.inspector = o.inspector
+	}
+	return nil
 }
 
-// Pack packs middlewares with the executor.
-func (h *Handler) Pack(e Executor) Executor {
+// pack packs middlewares with the executor.
+func (h *handler) pack(e Executor) (Executor, error) {
 	if e == nil {
-		return nil
+		return nil, NoExecutor.Error()
 	}
-	if len(h.Middlewares) <= 0 {
-		return e
+	if len(h.middlewares) <= 0 {
+		return e, nil
 	}
-	return NewMiddlewareExecutor(h.Middlewares, e)
+	return newMiddlewareExecutor(h.middlewares, e), nil
 }
 
-// UnionExecutor packs middlewares and own executor.
-func (h *Handler) UnionExecutor(ctx context.Context) Executor {
-	if h.Executor == nil {
-		return nil
+// unionExecutor packs middlewares and own executor.
+func (h *handler) unionExecutor(ctx context.Context) (Executor, error) {
```

Comment:
Still `Executor`?  Looks like you are using inspector here.

I don't have enough context about executor and inspector, can u give a high-level overview of the concepts?

Reply:
Inspector is used to generate executor. So  `unionExecutor()` is used to get an executor from 
 inspector.